---
title: "Simple simulation using VCF file"
author: "Marcin Kierczak"
date: "1/26/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library('vcfR')
library('stringr')
library('gwasim')
```

## Setup
Before starting our simulation, we need to load the necessary packages (install them earlier if you do not already have them in your library).
```{r setup_noexec, eval=FALSE}
library('vcfR')
library('stringr')
library('gwasim')
```

## Reading data
We will read first 500 rows (markers) from the *ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz* file that contains 1000 genomes phase 3 data for human chromosome 22.
```{r load_data}
input_vcf <- '~/Projects/johansson/ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz'
vcf_data <- vcfR::read.vcfR(file = input_vcf, nrows = 500)
markers <- vcf_data@fix[,'ID'] # Extract marker names
coords <- vcf_data@fix[, 'POS'] # Extract genomic marker coordinates
# Do some data cleaning
to_remove <- which(str_detect(markers, ";")) # detect weird markers (multi-marker)
markers <- markers[-to_remove]
coords <- coords[-to_remove]
```


## Get genotypes
```{r get_gt}
G <- gwasim::get_genotypes(x = vcf_data, marker_names = markers[1:10]) %>% 
  gwasim::fix_allele_encoding() %>%
  gwasim::impute_G()
```

## Compute maf

```{r compute_maf}
raf <- colSums(G) / dim(G)[1]
maf <- pmin(raf, 1-raf)
```

## Simulate effect of rare alleles for genotype
```{r simulate_effects}
effects <- gwasim::get_effects(maf=maf, thr=0.01, N=2, get_betas_fun = dbeta, get_betas_args = list(shape1 = .1, shape2 = .1), rare=T, frac_negative=0)
effects
```

## Compute simulated phenotypes
```{r compute_phenos}
error_term <- rnorm(n = dim(G)[1], mean = 0, sd = 1)
# Check the error is as you want it
plot(error_term)

# Simulate phenotypes
y <- G[, effects$marker_idx] %*% effects$effects + error_term
plot(y, pch = 1 + rowSums(G[, effects$marker_idx]))
```
