---
title: "Simulation more complex architectures in defined gene sets (vcf & bed input)"
author: "Marcin Kierczak"
date: "03/19/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#library('vcfR')
library('stringr')
library('gwasim')
library('reticulate')
library('SKAT')
```

## Synopsis
This document documents a bit more advanced example analyses of defined gene sets (bed file) in a whole genome (VCF file) using the `gwasim` package.
Package can be installed by typing: `devtools::install_github("NBISweden/a_johansson_2020")`. For further details visit [gwasim GitHub pages](https://github.com/NBISweden/a_johansson_2020).

## Setup
Before starting our simulation, we need to load the necessary packages (install them earlier if you do not already have them in your library). The easiest way is to:

* `install.packages('renv')`,
* download this Renv [lock file](https://github.com/NBISweden/a_johansson_2020/blob/master/renv.lock),
* `renv::restore()`

Next, load the packages:
```{r setup_noexec, eval=FALSE}
library('vcfR')
library('stringr')
library('data.table')
library('gwasim')
library('SKAT')
```

## Reading data

Because VCF files can be really large, we will have to emplot the windowing strategy, i.e. read a number of variants into memory, check if they are in any region of interest and discard the variants outside pre-defined regions.

### Read genic regions of interest from bed file

In this step, we will load gene regions of interest, in this case, promotor regions, and match them to the available genomic data. In this case, we will focus on only one chromosome, chr22. Data defining our regions is stored in a `.bed` file. We need to remember that bed files have their coordinates based at 0 while VCF files are 1-based (see, e.g. [details](https://www.biostars.org/p/84686/))
```{r read_bed}
bed_data <- read_delim("~/Projects/johansson/promoters.bed", delim = " ", col_names = c('chr', 'start', 'end', 'feature'))

# Convert to 1-based coords used by the VCF format, filter out unnecessary data
bed_data %<>% 
  mutate(start = start + 1) %>% 
  select(-feature) %>%
  filter(chr == "chr22")
```

## Read VCF in chunks
For the sake of learning how to use `gwasim` we will read first 500 rows (markers) from the `ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz` file. This file contains 1000 genomes phase 3 data for human chromosome 22. The file can be downloaded from the 1000 Genomes data repository via FTP [link](ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/).

> NOTE! For this part of the analysis, we need to setup Python whic we will use for improved efficiency of reading indexed VCF files in chunks.

```{r setup_python_noexec, eval = F}
reticulate::py_install('pandas', pip = T)
reticulate::py_install('pyvcf', pip = T)
```

```{r load_data, cache=TRUE}
path <- "/Users/kiero/Projects/johansson/ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz"
reader <- reticulate::import("vcf_region_reader")
N_sim <- 10
curr_sim <- 1
N_markers <- 5
results <- vector(mode = "list", length = N_sim)
while (curr_sim <= N_sim) {
  i <- sample(x = 1:dim(bed_data)[1], size = 1) # Select random region from bed
  region <- paste0(bed_data[i,'chr'], ':', bed_data[i,'start'], '-', bed_data[i,'end']) %>%
  str_remove("chr")
  data <- reader$vcf_region_reader(reticulate::r_to_py(path), region)
  G <- data %>% 
    gwasim::fix_allele_encoding() %>% 
    gwasim::impute_G()
  G <- as.matrix(apply(G, 2, as.integer))
  if (!('gwasim' %in% class(G))) class(G) <- c(class(G), 'gwasim')
  maf <- get_maf(G)
  effects <- gwasim::get_effects(maf=maf, thr=0.01, N=N_markers, get_betas_fun = dbeta, get_betas_args =
                                   list(shape1 = .1, shape2 = .1), rare=T, frac_negative=0, seed = 2)
  ## Filter 
  if (length(effects$marker_idx) >= N_markers) {
    error_term <- rnorm(n = dim(G)[1], mean = 0, sd = 1)
    y <- pmax(0, G[, effects$marker_idx] %*% effects$effects + error_term)
    results[[curr_sim]] <- list(marker_idx = effects$marker_idx, effects = effects$effects, y = y)
    null <- SKAT::SKAT_Null_Model(y ~ 1)
    skat <- SKAT::SKAT_CommonRare(G, obj = null)
    curr_sim <- curr_sim + 1
  }
}
```

