<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ gwasim</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gwasim</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="count_by_maf.html">count_by_maf()</a></code> </p>
        </td>
        <td><p>count number of loci below and above a maf threshold</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fix_allele_encoding.html">fix_allele_encoding()</a></code> </p>
        </td>
        <td><p>Fixes allele encoding so that the minor allele homozygote is always encoded as 2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_effects.html">get_effects()</a></code> </p>
        </td>
        <td><p>Get weights for markers using provided distribution function.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_genotypes.html">get_genotypes()</a></code> </p>
        </td>
        <td><p>Get genotypes matrix (counts of reference allele) from a GenABEL gwaa-data, vcfR object or
an object imported using seqminer</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_genotypes_gwaa_data.html">get_genotypes_gwaa_data()</a></code> </p>
        </td>
        <td><p>Get genotypes matrix (counts of reference allele) from a GenABEL gwaa-data object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_genotypes_seqminer_matrix.html">get_genotypes_seqminer_matrix()</a></code> </p>
        </td>
        <td><p>Get genotypes matrix (counts of reference allele) from one of the seqminer
read functions that return genotype matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_genotypes_vcfR.html">get_genotypes_vcfR()</a></code> </p>
        </td>
        <td><p>Get genotypes matrix (counts of reference allele) from a vcfR object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_maf.html">get_maf()</a></code> </p>
        </td>
        <td><p>Get minor allele frequency vector</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hush.html">hush()</a></code> </p>
        </td>
        <td><p>Silence too verbose code</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="impute_G.html">impute_G()</a></code> </p>
        </td>
        <td><p>Impute missing genotypes using minor allele frequency</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_effects.html">plot_effects()</a></code> </p>
        </td>
        <td><p>Plots effect sizes distribution dependent on minor allele frequency.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_mrk_density.html">plot_mrk_density()</a></code> </p>
        </td>
        <td><p>Plot density of markers in a region</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_region_vcf.html">read_region_vcf()</a></code> </p>
        </td>
        <td><p>read specific region from VCF file to a gwasim type genotype matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_regions_bed.html">read_regions_bed()</a></code> </p>
        </td>
        <td><p>Read regions information from bed file.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="recode_G.html">recode_G()</a></code> </p>
        </td>
        <td><p>Re-code genotypes according to the supplied genotype-phenotype map</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rnorm_betas.html">rnorm_betas()</a></code> </p>
        </td>
        <td><p>Effect sizes $$ from a normal distribution.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scan_CommonRare.html">scan_CommonRare()</a></code> </p>
        </td>
        <td><p>Run SKAT-O CommonRare model on all regions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sim_y_in_regions.html">sim_y_in_regions()</a></code> </p>
        </td>
        <td><p>Simulate phenotypes in a number of regions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="simulate_phenotype.html">simulate_phenotype()</a></code> <code><a href="simulate_phenotype.html">simulate_phenotype()</a></code> </p>
        </td>
        <td><p>Simulate phenotype given a set of common and rare variants.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tibble_to_gwaa.html">tibble_to_gwaa()</a></code> </p>
        </td>
        <td><p>Convert tibble with genotypes along with a vector of phenotypes to GenABEL internal raw gwaa.data format.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tibble_to_raw_genotypes.html">tibble_to_raw_genotypes()</a></code> </p>
        </td>
        <td><p>Convert tibble with genotypes to GenABEL internal raw genotype format and writ it into a file.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="validate_maf.html">validate_maf()</a></code> </p>
        </td>
        <td><p>Validate minor allele frequency vector</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


